Update Model Optimizer User Guide (#10759)
openvinotoolkit/openvino: docs/MO_DG/prepare_model/convert_model/Converting_Model.md

# Setting Input Shapes {#openvino_docs_MO_DG_prepare_model_convert_model_Converting_Model}

> Paragraphs below provide details about specifing input shapes for model conversion.

Author: Done

> ## When to Specify `--input_shape` Command Line Parameter <a name="when_to_specify_input_shapes"></a>

Reviewer: ```suggestion
## When to Specify `--input_shape` Command-line Parameter <a name="when_to_specify_input_shapes"></a>
```

Author: Done

> There are situations when Model Optimizer is unable to deduce input shapes of the model. In this case, user has to specify input shapes explicitly

Reviewer 2: This is partially out dated. We can freely work with undefined dimensions. And the problem should be sated as "we cannot always work with undefined input dimensions" and if you really run into this issue, you should specify it. Another reason is when we want to change the shape. Later should be references to dynamic shapes and reshape chapters in the runtime documentation. @jane-intel, please work with @rkazants to align on this stuff.

Author: Use static shapes when data size is determined and wouldn't be changed. So there will be no need to define it in run-time.
- User knows of data size and it is determined. So we can recommend to specify shapes for a model.
-

Author: "odel Optimizer is unable to deduce input shapes.." should also be mentioned

Author: Done

> There are situations when Model Optimizer is unable to deduce input shapes of the model. In this case, user has to specify input shapes explicitly

Author: Done

using `--input_shape` parameter. For example, run the Model Optimizer for the TensorFlow* MobileNet model with the single input
and specify input shape `[2,300,300,3]`.

```sh
mo --input_model MobileNet.pb --input_shape [2,300,300,3]
```

> If a model has multiple inputs, `--input_shape` must be used in conjunction with `--input` parameter.

Author: Due to limitation you mentioned let us not describe the syntax of `--input` for specifying shapes. Now we have the smooth story flow from `--input_shape` to `--input`, from one input to multiple inputs.

The parameter `--input` contains a list of input names for which shapes in the same order are defined via `--input_shape`.
For example, launch the Model Optimizer for the ONNX* OCR model with a pair of inputs `data` and `seq_len` 
and specify shapes `[3,150,200,1]` and `[3]` for them.

```sh
mo --input_model ocr.onnx --input data,seq_len --input_shape [3,150,200,1],[3]
```

> The parameter `--input_shape` allows to override original input shapes to ones compatible for a given model.

Reviewer: ```suggestion
The parameter `--input_shape` allows overriding original input shapes to the shapes compatible with a given model.
```

Author: Done

> Dynamic shapes, i.e. with dynamic dimensions, in the original model can be replaced with static shapes for the converted model, and vice versa.

Reviewer: This is kinda unclear 
Dynamic shapes, i.e. with dynamic dimensions, in the original model can be replaced with static shapes for the converted model, and vice versa.
Does it mean - Dynamic shapes that have dynamic dimensions in the original model can be replaced with static shapes in the converted model, and vice versa.

Author: yes. Please clarify How to rephrase better.

The dynamic dimension can be marked in Model Optimizer command-line as `-1` or `?`.
For example, launch the Model Optimizer for the ONNX* OCR model and specify dynamic batch dimension for inputs.

```sh
mo --input_model ocr.onnx --input data,seq_len --input_shape [-1,150,200,1],[-1]
```

## When to Specify `--static_shape` Command Line Parameter
If the `--static_shape` command line parameter is specified, the Model Optimizer evaluates shapes of all operations in the model (shape propagation)
for a fixed input(s) shape(s). During the shape propagation the Model Optimizer evaluates operations *Shape* and removes them from the computation graph.
With that approach, the initial model which can consume inputs of different shapes may be converted to IR working with the input of one fixed shape only.
For example, consider the case when some blob is reshaped from 4D of a shape *[N, C, H, W]* to a shape *[N, C, H \* W]*.
During the model conversion the Model Optimize calculates output shape as a constant 1D blob with values *[N, C, H \* W]*.
So if the input shape changes to some other value *[N, C, H1, W1]* (it is possible scenario for a fully convolutional model) then the reshape layer

> becomes invalid. Resulting Intermediate Representation will not be resizable with the help of OpenVINO Runtime API.

Reviewer 2: Too low-level description. Example is cryptic. Please rewrite the whole thing focusing on: use case when it provides value (smaller IR and shorted load time), and limitation -- inability to change shape later in run-time. Describe why just --input_shape is not enough.

Author: Use when data size is determined and would be changed. User can freeze shapes of a model.
Values:
 - compact model
 - improved loading time

Author: Done

## See Also
* [Introduction](../../Deep_Learning_Model_Optimizer_DevGuide.md)
* [Cutting Off Parts of a Model](Cutting_Model.md)
* [Optimizing Preprocessing Computation](../Additional_Optimizations.md)
* [Compression of a Model to FP16](../FP16_Compression.md)
* [Converting TensorFlow Models](Convert_Model_From_TensorFlow.md)
* [Converting ONNX Models](Convert_Model_From_ONNX.md)
* [Converting PyTorch Models](Convert_Model_From_PyTorch.md)
* [Converting PaddlePaddle Models](Convert_Model_From_Paddle.md)
* [Converting MXNet Models](Convert_Model_From_MxNet.md)
* [Converting Caffe Models](Convert_Model_From_Caffe.md)

> * [Converting Kaldi Models](Convert_Model_From_Kaldi.md)

Reviewer 3: do we need so long list? I suppose the page is renamed from conversion steps to input shapes steps, so `See also` can refer to `Changing input shapes` in runtime `openvino_docs_IE_DG_ShapeInference.html`

Author: Corrected "See also" list. Now we can't easily change these filenames because URLs will be broken. We need to make aliases probably for existing names. @andrew-zaytsev, please advise - can we rename doc files?

Sometimes Model Optimizer is unable to convert models out-of-the-box (only the `--input_model` parameter is specified).
Such problem can relate to models with inputs of undefined ranks and a case of cutting off parts of a model.
In this case, user has to specify input shapes explicitly using `--input_shape` parameter.

For example, run the Model Optimizer for the TensorFlow* MobileNet model with the single input
and specify input shape `[2,300,300,3]`.

```sh
mo --input_model MobileNet.pb --input_shape [2,300,300,3]
```

If a model has multiple inputs, `--input_shape` must be used in conjunction with `--input` parameter.
The parameter `--input` contains a list of input names for which shapes in the same order are defined via `--input_shape`.
For example, launch the Model Optimizer for the ONNX* OCR model with a pair of inputs `data` and `seq_len` 
and specify shapes `[3,150,200,1]` and `[3]` for them.

```sh
mo --input_model ocr.onnx --input data,seq_len --input_shape [3,150,200,1],[3]
```

The alternative way to specify input shapes is to use the `--input` parameter as follows:

```sh
mo --input_model ocr.onnx --input data[3 150 200 1],seq_len[3]
```

> The parameter `--input_shape` allows overriding original input shapes to the shapes compatible for a given model.

Reviewer: ```suggestion
The parameter `--input_shape` allows overriding original input shapes to the shapes compatible with a given model.
```

Dynamic shapes, i.e. with dynamic dimensions, in the original model can be replaced with static shapes for the converted model, and vice versa.
The dynamic dimension can be marked in Model Optimizer command-line as `-1` or `?`.
For example, launch the Model Optimizer for the ONNX* OCR model and specify dynamic batch dimension for inputs.

```sh
mo --input_model ocr.onnx --input data,seq_len --input_shape [-1,150,200,1],[-1]
```

> To optimize memory consumption for models with undefined dimensions in run-time, Model Optimizer provides capability to define boundaries of dimensions.

Reviewer: ```suggestion
To optimize memory consumption for models with undefined dimensions in run-time, Model Optimizer provides the capability to define boundaries of dimensions.
```

The boundaries of undefined dimension can be specified with ellipsis.
For example, launch the Model Optimizer for the ONNX* OCR model and specify a boundary for the batch dimension.

```sh
mo --input_model ocr.onnx --input data,seq_len --input_shape [1..3,150,200,1],[1..3]
```

## When to Specify --static_shape Command-line Parameter

> Model Optimizer provides the `--static_shape` parameter that allows to evaluate shapes of all operations in the model for fixed input shapes

Reviewer: ```suggestion
Model Optimizer provides the `--static_shape` parameter that allows evaluating shapes of all operations in the model for fixed input shapes
```

> and to fold shape computing sub-graphs into constants. The resulted IR can be more compact in size and the loading time for such IR can be decreased.
> However, the resulted IR will not be reshape-able with the help of the `reshape` method from OpenVINO Runtime API.
> It is worth to note that the `--input_shape` parameter does not affect reshape-ability of the model.

Reviewer: ```suggestion
and to fold shape computing sub-graphs into constants. The resulting IR can be more compact in size and the loading time for such IR can be decreased.
However, the resulting IR will not be reshape-able with the help of the `reshape` method from OpenVINO Runtime API.
It is worth noting that the `--input_shape` parameter does not affect reshape-ability of the model.
```

* [Introduction](../../Deep_Learning_Model_Optimizer_DevGuide.md)
* [Cutting Off Parts of a Model](Cutting_Model.md)
* [Optimizing Preprocessing Computation](../Additional_Optimizations.md)

> * [Compression Of Model To FP16](FP16_Compression.md)

Author: `To` -> `to`
